src/include/art.h

```diff
@@ -0,0 +1,257 @@
+};
+
+struct art_iterator
+{
+   struct deque* que;
+   struct art* tree;
+   uint32_t count;
+};
+
+/**
+ * Initializes an adaptive radix tree
+ * @param tree [out] The tree
+ * @param val_destroy_cb The callback to destroy the val in leaf,
+ * simple free() is used as default if input is NULL.
+ * See pgmoneta_art_destroy_value_noop() if you don't need value to be freed.
+ * @return 0 on success
+ */
+int
+pgmoneta_art_init(struct art** tree, value_destroy_callback val_destroy_cb);
+
+/**
+ * Destroys an ART tree
+ * @return 0 on success
+ */
+int
+pgmoneta_art_destroy(struct art* tree);
+
+/**
+ * inserts a new value into the art tree, note that the key is copied while the value is not
+ * @param t the tree
+ * @param key the key
+ * @param key_len the length of the key
+ * @param value opaque value
+ * @return null if the item was newly inserted, otherwise
+ * the old value pointer is returned
+ */
+void*
+pgmoneta_art_insert(struct art* t, unsigned char* key, int key_len, void* value);
+
+/**
+ * Deletes a value from the ART tree
+ * @param t The tree
+ * @param key The key
+ * @param key_len The length of the key
+ * @return NULL if the item was not found, otherwise
+ * the value pointer is returned
+ */
+void*
+pgmoneta_art_delete(struct art* t, unsigned char* key, int key_len);
+
+/**
+ * Searches for a value in the ART tree
+ * @param t The tree
+ * @param key The key
+ * @param key_len The length of the key
+ * @return NULL if the item was not found, otherwise
+ * the value pointer is returned
+ */
+void*
+pgmoneta_art_search(struct art* t, unsigned char* key, int key_len);
+
+/**
+ * Iterates through the entries pairs in the map,
+ * invoking a callback for each. The call back gets a
+ * key, value for each and returns an integer stop value
+ * If the callback returns non-zero, then the iteration stops
+ * @param t The tree to iterate over
+ * @param cb The callback function to invoke
+ * @param data Opaque handle passed to the callback
+ * @return 0 on success, or the return of the callback
+ */
+int
+pgmoneta_art_iterate(struct art* t, art_callback cb, void* data);
+
+/**
+ * Create an art iterator
+ * @param iter [out] The iterator
+ * @param t The tree
+ * @return 0 if success, otherwise 1
+ */
+int
+pgmoneta_art_iterator_init(struct art_iterator** iter, struct art* t);
+
+/**
+ * Destroy the iterator
+ * @param iter The iterator
+ */
+void
+pgmoneta_art_iterator_destroy(struct art_iterator* iter);
+
+/**
+ * Return the next leaf in the ART
+ * @param iter The iterator
+ * @return The next leaf, or NULL if all leaves have been visited
+ */
+struct art_leaf*
+pgmoneta_art_iterator_next(struct art_iterator* iter);
+
+/**
+ * Check if iterator has next
+ * @param iter
+ * @return
+ */
+bool
+pgmoneta_art_iterator_has_next(struct art_iterator* iter);
+
+/**
+ * The noop callback function for destroying value when destroying ART
+ * @param val The value
+ */
+void
+pgmoneta_art_destroy_value_noop(void* val);
```

Comment:
Do we need this ?

Reply:
We don't now. I put it just in case in the future we don't want to free the value... And we'll then have an noop ready to use.